### PR TITLE
Save plotly figure

### PR DESCRIPTION
**Pull request by tmichela**
Allow saving plotly figures as a variable result.
GUI widget will be as a separate PR.

### CODE REVIEW

File: damnit/ctxsupport/ctxrunner.py [hunk 1]
```diff
@@ -14,24 +14,23 @@
...
 import numpy as np
-import xarray as xr
 import requests
+import xarray as xr
 import yaml
-
 from damnit_ctx import RunData, Variable
+from matplotlib.axes import Axes
+from matplotlib.figure import Figure
+from plotly.graph_objects import Figure as PlotlyFigure
```
Comment (JamesWrigley) [inline damnit/ctxsupport/ctxrunner.py:33]: We need to document the new dependency: https://damnit.readthedocs.io/en/latest/backend/#using-custom-environments

**Review by JamesWrigley — CHANGES_REQUESTED** [review]
Should we do this automagically for all Figure's? I'm inclined to say yes, but let's see how it works first :octopus: Also, how large are the datasets that are produced? I imagine they can get into multiple megabytes pretty easily, is it worth compressing the JSON first?

File: damnit/ctxsupport/ctxrunner.py (hunk 1 above)
Comment (takluyver) [reply damnit/ctxsupport/ctxrunner.py:34]: Alternatively, we could do the check without making plotly a dependency, using a version of the `isinstance_no_import` function we've got in a couple of places already:

https://github.com/European-XFEL/EXtra/blob/7d0f66f781321f84b73790f27ccba11854a7608a/src/extra/components/utils.py#L30-L36

(I got this technique from IPython)

**Comment by tmichela** [comment]
> Should we do this automagically for all Figure's?

What do you mean by "all Figure"? 🤔 

> Also, how large are the datasets that are saved?  I imagine they can get into multiple megabytes pretty easily, is it worth compressing the JSON first?

About 3MB for XGM data example. A quick test gives me about 50% compression on that data.

**Comment by JamesWrigley** [comment]
> What do you mean by "all Figure"? 🤔

As in, converting regular matplotlib figures to plotly figures by default. I'm assuming that's possible from the XGM example? :eyes: 

> About 3MB for XGM data example. A quick test gives me about 50% compression on that data (using gzip).

Ok, yeah I think I'd go for that then.

File: damnit/api.py [hunk 2]
```diff
@@ -19,6 +20,7 @@ class DataType(Enum):
     Dataset = "dataset"
     Image = "image"
     Timestamp = "timestamp"
+    PlotlyFigure = "PlotlyFigure"
```
Comment (CammilleCC) [inline damnit/api.py:23]: Nitpick: `PlotlyJson = "plotly_json` perhaps?

**Review by CammilleCC — COMMENTED** [review]
LGTM, thanks for the addition!

**Comment by tmichela** [comment]
> As in, converting regular matplotlib figures to plotly figures by default. I'm assuming that's possible from the XGM example? 👀

huh, no I don't think that's possible. I just re-created the plots for comparison.

**Comment by JamesWrigley** [comment]
Ahh, devastating...

File: damnit/api.py (hunk 2 above)
Comment (tmichela) [reply damnit/api.py:23]: I prefer `PlotlyFigure` as this is the object that is to be saved, and returned by the API.

**Comment by tmichela** [comment]
> Ok, yeah I think I'd go for that then.

Compression is implemented. Not sure it's super worth it but 🤷

**Review by JamesWrigley — APPROVED** [review]
LGTM!

**Comment by tmichela** [comment]
Thanks!

**Pull request merged**